common/src/reserved.rs

```diff
@@ -59,18 +78,79 @@ impl ReservedState {
+        }
+        result.sort_by(|a, b| b.0.cmp(&a.0));
+        Ok(result)
     }
 
-    pub fn apply_delegate(&mut self, _tx: &TxDelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_delegate(&mut self, tx: &TxDelegate) -> Result<Self, String> {
+        let mut delegator_updated = false;
+        let delegator_name = match self.query_name(&tx.delegator.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegator does not exist by name".to_string()),
+        };
+        let delegatee_name = match self.query_name(&tx.delegatee.clone()) {
+            Some(name) => name,
+            None => return Result::Err("delegatee does not exist by name".to_string()),
+        };
+        for delegator in &mut self.members {
+            if delegator.name == delegator_name {
+                if tx.delegatee.to_string().is_empty() {
+                    return Result::Err("delegatee field cannot be empty".to_string());
+                }
+                if tx.governance {
+                    delegator.governance_delegatee = Option::from(delegatee_name.clone());
+                    delegator.consensus_delegatee = Option::from(delegatee_name.clone());
+                    delegator_updated = true;
+                } else {
+                    delegator.consensus_delegatee = Option::from(delegatee_name.clone());
+                    delegator_updated = true;
+                }
+            }
+        }
+        if delegator_updated {
+            Ok(self.clone())
+        } else {
+            Err("nothing was delegated. please check your transaction again.".to_string())
+        }
     }
 
-    pub fn apply_undelegate(&mut self, _tx: &TxUndelegate) -> Result<Self, String> {
-        unimplemented!()
+    pub fn apply_undelegate(&mut self, tx: &TxUndelegate) -> Result<Self, String> {
+        let delegator_name = match self.query_name(&tx.delegator.clone()) {
+            Some(name) => name,
+            None => return Err("delegator does not exist by name".to_string()),
+        };
+        let mut delegator_updated = false;
+        for delegator in &mut self.members {
+            if delegator.name == delegator_name {
+                if let Some(_consensus_delegatee) = &delegator.consensus_delegatee {
+                    delegator.consensus_delegatee = None;
+                    if tx.governance {
+                        if let Some(_governance_delegatee) = &delegator.governance_delegatee {
+                            delegator.governance_delegatee = None;
+                            delegator_updated = true;
+                        } else {
+                            return Err("governance delegatee is not set".to_string());
+                        }
+                    } else {
+                        delegator_updated = true;
+                    }
+                } else {
+                    return Err("consensus delegatee is not set".to_string());
+                }
```

Comment:
Ditto.

Reply:
same - can this happen when the transaction doesn't hold the existing delegatee in some reasons?